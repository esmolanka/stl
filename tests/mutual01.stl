mutual
  | Even = < EZero, ESucc : Odd >
  | Odd  = < OZero, OSucc : Even >

#eval Even

#eval Odd

#check <EZero, ESucc : Odd> <: Even
#check Even <: <EZero, ESucc : Odd>

#check <OZero, OSucc : Even> <: Odd
#check Odd <: <OZero, OSucc : Even>
