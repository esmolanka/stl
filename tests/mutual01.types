type Even = < EZero | ESucc : Odd >
  with Odd = < OZero | OSucc : Even >

#eval Even

#eval Odd

#check <EZero | ESucc : Odd> <: Even
#check Even <: <EZero | ESucc : Odd>

#check <OZero | OSucc : Even> <: Odd
#check Odd <: <OZero | OSucc : Even>
