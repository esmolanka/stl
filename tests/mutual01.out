tests/mutual01.stl:4:1-4:11:
  Normalised:
    μ Even. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < EZero^ω : Unit
      | ESucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <OZero^ω : Unit| OSucc^ω/1 : Even | ... ρ>
      | ... ρ
      >
  Kind:
    Type
tests/mutual01.stl:6:1-6:10:
  Normalised:
    μ Odd. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < OZero^ω : Unit
      | OSucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <EZero^ω : Unit| ESucc^ω/1 : Odd | ... ρ>
      | ... ρ
      >
  Kind:
    Type
tests/mutual01.stl:8:1-8:37:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < EZero^ω : Unit
      | ESucc^ω/1 :
          μ Odd. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            < OZero^ω : Unit
            | OSucc^ω/1 :
                ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <EZero^ω : Unit| ESucc^ω/1 : Odd | ... ρ>
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Even. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < EZero^ω : Unit
      | ESucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <OZero^ω : Unit| OSucc^ω/1 : Even | ... ρ>
      | ... ρ
      >
  Result:
    OK
tests/mutual01.stl:9:1-9:37:
  Subtype:
    μ Even. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < EZero^ω : Unit
      | ESucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <OZero^ω : Unit| OSucc^ω/1 : Even | ... ρ>
      | ... ρ
      >
  Supertype:
    ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < EZero^ω : Unit
      | ESucc^ω/1 :
          μ Odd. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            < OZero^ω : Unit
            | OSucc^ω/1 :
                ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <EZero^ω : Unit| ESucc^ω/1 : Odd | ... ρ>
            | ... ρ
            >
      | ... ρ
      >
  Result:
    OK
tests/mutual01.stl:11:1-11:37:
  Subtype:
    ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < OZero^ω : Unit
      | OSucc^ω/1 :
          μ Even. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            < EZero^ω : Unit
            | ESucc^ω/1 :
                ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <OZero^ω : Unit| OSucc^ω/1 : Even | ... ρ>
            | ... ρ
            >
      | ... ρ
      >
  Supertype:
    μ Odd. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < OZero^ω : Unit
      | OSucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <EZero^ω : Unit| ESucc^ω/1 : Odd | ... ρ>
      | ... ρ
      >
  Result:
    OK
tests/mutual01.stl:12:1-12:37:
  Subtype:
    μ Odd. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < OZero^ω : Unit
      | OSucc^ω/1 :
          ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            <EZero^ω : Unit| ESucc^ω/1 : Odd | ... ρ>
      | ... ρ
      >
  Supertype:
    ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
      < OZero^ω : Unit
      | OSucc^ω/1 :
          μ Even. ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
            < EZero^ω : Unit
            | ESucc^ω/1 :
                ∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
                  <OZero^ω : Unit| OSucc^ω/1 : Even | ... ρ>
            | ... ρ
            >
      | ... ρ
      >
  Result:
    OK